### Deep Learning(keras)

1. Deep Learning이란?
    - machine lenarning의 일종.<br>
    데이터를 통해 컴퓨터가 '학습'을 하고 그것을 사용해 컴퓨터가 자동으로 문제를 해결하도록 하는 것.<br>
    이 때 프로그램에 적용하는 rule은 원래 사람이 만들어서 줬었음.<br>
    그런데 사람이 rule을 만들 때에는 어려움이 존재하기 때문에 data와 정답으로 기계를 학습시켜(machine learning) rule을 만들어 프로그램에 적용시킴.<br>
        - high level feature: 사람의 처리를 거쳐 다루기 쉬운 데이터.
        - low level feature: 그 반대.
    - Perceptron: 뇌의 신경세포를 모델링해서 만든 것.그러나 문제 많았음.<br>
    -> Hidden layer 추가:  학습 정말 잘 됨. 그러나 사람이 이해할 수 없음.
    - deep learning이 다시 뜬 이유
        - 하드웨어의 발전
        - 빅 데이터(데이터의 수가 많아짐)
        - 기존 문제 해결(Gradient Vanishment, Overfitting)
    

2. FFNN Model
    - 뉴럴 네트워크의 output에서 calculate error를 계산해 다시 input으로?!

In [2]:
import pandas as pd

In [3]:
path = 'C:/Download/train.csv'

In [4]:
titanic = pd.read_csv(path)

In [5]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

In [6]:
titanic = titanic.drop(columns = ['PassengerId', 'Name', 'Ticket', 'Cabin'])

In [7]:
titanic

Survived  Pclass     Sex   Age  SibSp  Parch      Fare Embarked
0           0       3    male  22.0      1      0    7.2500        S
1           1       1  female  38.0      1      0   71.2833        C
2           1       3  female  26.0      0      0    7.9250        S
3           1       1  female  35.0      1      0   53.1000        S
4           0       3    male  35.0      0      0    8.0500        S
5           0       3    male   NaN      0      0    8.4583        Q
6           0       1    male  54.0      0      0   51.8625        S
7           0       3    male   2.0      3      1   21.0750        S
8           1       3  female  27.0      0      2   11.1333        S
9           1       2  female  14.0      1      0   30.0708        C
10          1       3  female   4.0      1      1   16.7000        S
11          1       1  female  58.0      0      0   26.5500        S
12          0       3    male  20.0      0      0    8.0500        S
13          0       3    male  39.0      1      5   31.2750        S
14          0       3  female  14.0      0      0    7.8542        S
15          1       2  female  55.0      0      0   16.0000        S
16          0       3    male   2.0      4      1   29.1250        Q
17          1       2    male   NaN      0      0   13.0000        S
18          0       3  female  31.0      1      0   18.0000        S
19          1       3  female   NaN      0      0    7.2250        C
20          0       2    male  35.0      0      0   26.0000        S
21          1       2    male  34.0      0      0   13.0000        S
22          1       3  female  15.0      0      0    8.0292        Q
23          1       1    male  28.0      0      0   35.5000        S
24          0       3  female   8.0      3      1   21.0750        S
25          1       3  female  38.0      1      5   31.3875        S
26          0       3    male   NaN      0      0    7.2250        C
27          0       1    male  19.0      3      2  263.0000        S
28          1       3  female   NaN      0      0    7.8792        Q
29          0       3    male   NaN      0      0    7.8958        S
..        ...     ...     ...   ...    ...    ...       ...      ...
861         0       2    male  21.0      1      0   11.5000        S
862         1       1  female  48.0      0      0   25.9292        S
863         0       3  female   NaN      8      2   69.5500        S
864         0       2    male  24.0      0      0   13.0000        S
865         1       2  female  42.0      0      0   13.0000        S
866         1       2  female  27.0      1      0   13.8583        C
867         0       1    male  31.0      0      0   50.4958        S
868         0       3    male   NaN      0      0    9.5000        S
869         1       3    male   4.0      1      1   11.1333        S
870         0       3    male  26.0      0      0    7.8958        S
871         1       1  female  47.0      1      1   52.5542        S
872         0       1    male  33.0      0      0    5.0000        S
873         0       3    male  47.0      0      0    9.0000        S
874         1       2  female  28.0      1      0   24.0000        C
875         1       3  female  15.0      0      0    7.2250        C
876         0       3    male  20.0      0      0    9.8458        S
877         0       3    male  19.0      0      0    7.8958        S
878         0       3    male   NaN      0      0    7.8958        S
879         1       1  female  56.0      0      1   83.1583        C
880         1       2  female  25.0      0      1   26.0000        S
881         0       3    male  33.0      0      0    7.8958        S
882         0       3  female  22.0      0      0   10.5167        S
883         0       2    male  28.0      0      0   10.5000        S
884         0       3    male  25.0      0      0    7.0500        S
885         0       3  female  39.0      0      5   29.1250        Q
886         0       2    male  27.0      0      0   13.0000        S
887

In [8]:
titanic.count()

Survived    891
Pclass      891
Sex         891
Age         714
SibSp       891
Parch       891
Fare        891
Embarked    889
dtype: int64

Deep leanring은 최대한 dataset의 수가 많으면 좋기 때문에 null값을 다른 값으로 채워주자.

In [9]:
from sklearn.impute import SimpleImputer

In [10]:
imputer = SimpleImputer(strategy='most_frequent')

In [11]:
titanic[['Embarked']] = imputer.fit_transform(titanic[['Embarked']])

Embarked의 NULL값이 채워짐.

In [12]:
titanic.count()

Survived    891
Pclass      891
Sex         891
Age         714
SibSp       891
Parch       891
Fare        891
Embarked    891
dtype: int64

In [13]:
imputer.set_params(strategy='mean')

SimpleImputer(copy=True, fill_value=None, missing_values=nan, strategy='mean',
       verbose=0)

In [14]:
titanic[['Age']] = imputer.fit_transform(titanic[['Age']])

Age의 NULL값도 채워져서 비어있던 정보들이 전부 채워짐.<br>
Embarked의 경우에는 가장 많이 나타는 글자로 대체한 것이고, Age의 경우에는 전체 나이의 평균 값을 넣어 줌.

In [15]:
titanic.count()

Survived    891
Pclass      891
Sex         891
Age         891
SibSp       891
Parch       891
Fare        891
Embarked    891
dtype: int64

In [16]:
titanic

Survived  Pclass     Sex        Age  SibSp  Parch      Fare Embarked
0           0       3    male  22.000000      1      0    7.2500        S
1           1       1  female  38.000000      1      0   71.2833        C
2           1       3  female  26.000000      0      0    7.9250        S
3           1       1  female  35.000000      1      0   53.1000        S
4           0       3    male  35.000000      0      0    8.0500        S
5           0       3    male  29.699118      0      0    8.4583        Q
6           0       1    male  54.000000      0      0   51.8625        S
7           0       3    male   2.000000      3      1   21.0750        S
8           1       3  female  27.000000      0      2   11.1333        S
9           1       2  female  14.000000      1      0   30.0708        C
10          1       3  female   4.000000      1      1   16.7000        S
11          1       1  female  58.000000      0      0   26.5500        S
12          0       3    male  20.000000      0      0    8.0500        S
13          0       3    male  39.000000      1      5   31.2750        S
14          0       3  female  14.000000      0      0    7.8542        S
15          1       2  female  55.000000      0      0   16.0000        S
16          0       3    male   2.000000      4      1   29.1250        Q
17          1       2    male  29.699118      0      0   13.0000        S
18          0       3  female  31.000000      1      0   18.0000        S
19          1       3  female  29.699118      0      0    7.2250        C
20          0       2    male  35.000000      0      0   26.0000        S
21          1       2    male  34.000000      0      0   13.0000        S
22          1       3  female  15.000000      0      0    8.0292        Q
23          1       1    male  28.000000      0      0   35.5000        S
24          0       3  female   8.000000      3      1   21.0750        S
25          1       3  female  38.000000      1      5   31.3875        S
26          0       3    male  29.699118      0      0    7.2250        C
27          0       1    male  19.000000      3      2  263.0000        S
28          1       3  female  29.699118      0      0    7.8792        Q
29          0       3    male  29.699118      0      0    7.8958        S
..        ...     ...     ...        ...    ...    ...       ...      ...
861         0       2    male  21.000000      1      0   11.5000        S
862         1       1  female  48.000000      0      0   25.9292        S
863         0       3  female  29.699118      8      2   69.5500        S
864         0       2    male  24.000000      0      0   13.0000        S
865         1       2  female  42.000000      0      0   13.0000        S
866         1       2  female  27.000000      1      0   13.8583        C
867         0       1    male  31.000000      0      0   50.4958        S
868         0       3    male  29.699118      0      0    9.5000        S
869         1       3    male   4.000000      1      1   11.1333        S
870         0       3    male  26.000000      0      0    7.8958        S
871         1       1  female  47.000000      1      1   52.5542        S
872         0       1    male  33.000000      0      0    5.0000        S
873         0       3    male  47.000000      0      0    9.0000        S
874         1       2  female  28.000000      1      0   24.0000        C
875         1       3  female  15.000000      0      0    7.2250        C
876         0       3    male  20.000000      0      0    9.8458        S
877         0       3    male  19.000000      0      0    7.8958        S
878         0       3    male  29.699118      0      0    7.8958        S
879         1       1  female  56.000000      0      1   83.1583        C
880         1       2  female  25.000000      0      1   26.0000        S
881         0       3    male  33.000000      0      0    7.8958        S
882         0       3  female  22.000000      0      0   10.5167        S
883      

pandas에도 OneHotEncoder가 있음!

In [17]:
sex_df = pd.get_dummies(titanic[['Sex']])

In [18]:
sex_df

Sex_female  Sex_male
0             0         1
1             1         0
2             1         0
3             1         0
4             0         1
5             0         1
6             0         1
7             0         1
8             1         0
9             1         0
10            1         0
11            1         0
12            0         1
13            0         1
14            1         0
15            1         0
16            0         1
17            0         1
18            1         0
19            1         0
20            0         1
21            0         1
22            1         0
23            0         1
24            1         0
25            1         0
26            0         1
27            0         1
28            1         0
29            0         1
..          ...       ...
861           0         1
862           1         0
863           1         0
864           0         1
865           1         0
866           1         0
867           0         1
868           0         1
869           0         1
870           0         1
871           1         0
872           0         1
873           0         1
874           1         0
875           1         0
876           0         1
877           0         1
878           0         1
879           1         0
880           1         0
881           0         1
882           1         0
883           0         1
884           0         1
885           1         0
886           0         1
887           1         0
888           1         0
889           0         1
890           0         1

[891 rows x 2 columns]

In [19]:
emb_df = pd.get_dummies(titanic[['Embarked']])

In [20]:
emb_df

Embarked_C  Embarked_Q  Embarked_S
0             0           0           1
1             1           0           0
2             0           0           1
3             0           0           1
4             0           0           1
5             0           1           0
6             0           0           1
7             0           0           1
8             0           0           1
9             1           0           0
10            0           0           1
11            0           0           1
12            0           0           1
13            0           0           1
14            0           0           1
15            0           0           1
16            0           1           0
17            0           0           1
18            0           0           1
19            1           0           0
20            0           0           1
21            0           0           1
22            0           1           0
23            0           0           1
24            0           0           1
25            0           0           1
26            1           0           0
27            0           0           1
28            0           1           0
29            0           0           1
..          ...         ...         ...
861           0           0           1
862           0           0           1
863           0           0           1
864           0           0           1
865           0           0           1
866           1           0           0
867           0           0           1
868           0           0           1
869           0           0           1
870           0           0           1
871           0           0           1
872           0           0           1
873           0           0           1
874           1           0           0
875           1           0           0
876           0           0           1
877           0           0           1
878           0           0           1
879           1           0           0
880           0           0           1
881           0           0           1
882           0           0           1
883           0           0           1
884           0           0           1
885           0           1           0
886           0           0           1
887           0           0           1
888           0           0           1
889           1           0           0
890           0           1           0

[891 rows x 3 columns]

In [21]:
titanic = pd.concat((titanic.drop(columns=['Sex', 'Embarked']), sex_df, emb_df), axis=1)

In [22]:
titanic

Survived  Pclass        Age  SibSp  Parch      Fare  Sex_female  \
0           0       3  22.000000      1      0    7.2500           0   
1           1       1  38.000000      1      0   71.2833           1   
2           1       3  26.000000      0      0    7.9250           1   
3           1       1  35.000000      1      0   53.1000           1   
4           0       3  35.000000      0      0    8.0500           0   
5           0       3  29.699118      0      0    8.4583           0   
6           0       1  54.000000      0      0   51.8625           0   
7           0       3   2.000000      3      1   21.0750           0   
8           1       3  27.000000      0      2   11.1333           1   
9           1       2  14.000000      1      0   30.0708           1   
10          1       3   4.000000      1      1   16.7000           1   
11          1       1  58.000000      0      0   26.5500           1   
12          0       3  20.000000      0      0    8.0500           0   
13          0       3  39.000000      1      5   31.2750           0   
14          0       3  14.000000      0      0    7.8542           1   
15          1       2  55.000000      0      0   16.0000           1   
16          0       3   2.000000      4      1   29.1250           0   
17          1       2  29.699118      0      0   13.0000           0   
18          0       3  31.000000      1      0   18.0000           1   
19          1       3  29.699118      0      0    7.2250           1   
20          0       2  35.000000      0      0   26.0000           0   
21          1       2  34.000000      0      0   13.0000           0   
22          1       3  15.000000      0      0    8.0292           1   
23          1       1  28.000000      0      0   35.5000           0   
24          0       3   8.000000      3      1   21.0750           1   
25          1       3  38.000000      1      5   31.3875           1   
26          0       3  29.699118      0      0    7.2250           0   
27          0       1  19.000000      3      2  263.0000           0   
28          1       3  29.699118      0      0    7.8792           1   
29          0       3  29.699118      0      0    7.8958           0   
..        ...     ...        ...    ...    ...       ...         ...   
861         0       2  21.000000      1      0   11.5000           0   
862         1       1  48.000000      0      0   25.9292           1   
863         0       3  29.699118      8      2   69.5500           1   
864         0       2  24.000000      0      0   13.0000           0   
865         1       2  42.000000      0      0   13.0000           1   
866         1       2  27.000000      1      0   13.8583           1   
867         0       1  31.000000      0      0   50.4958           0   
868         0       3  29.699118      0      0    9.5000           0   
869         1       3   4.000000      1      1   11.1333           0   
870         0       3  26.000000      0      0    7.8958           0   
871         1       1  47.000000      1      1   52.5542           1   
872         0       1  33.000000      0      0    5.0000           0   
873         0       3  47.000000      0      0    9.0000           0   
874         1       2  28.000000      1      0   24.0000           1   
875         1       3  15.000000      0      0    7.2250           1   
876         0       3  20.000000      0      0    9.8458           0   
877         0       3  19.000000      0      0    7.8958           0   
878         0       3  29.699118      0      0    7.8958           0   
879         1       1  56.000000      0      1   83.1583           1   
880         1       2  25.000000      0      1   26.0000           1   
881         0       3  33.000000      0      0    7.8958           0   
882         0       3  22.000000      0      0   10.5167           1   
883         0       2  28.000000      0      0   10.5000           0   
884         0       3  25.000000      0      

In [23]:
train = titanic.sample(frac=0.8)
valid = titanic.drop(index=train.index)

In [24]:
train

Survived  Pclass        Age  SibSp  Parch      Fare  Sex_female  \
863         0       3  29.699118      8      2   69.5500           1   
108         0       3  38.000000      0      0    7.8958           0   
486         1       1  35.000000      1      0   90.0000           1   
700         1       1  18.000000      1      0  227.5250           1   
608         1       2  22.000000      1      2   41.5792           1   
184         1       3   4.000000      0      2   22.0250           1   
615         1       2  24.000000      1      2   65.0000           1   
216         1       3  27.000000      0      0    7.9250           1   
659         0       1  58.000000      0      2  113.2750           0   
682         0       3  20.000000      0      0    9.2250           0   
471         0       3  38.000000      0      0    8.6625           0   
536         0       1  45.000000      0      0   26.5500           0   
566         0       3  19.000000      0      0    7.8958           0   
829         1       1  62.000000      0      0   80.0000           1   
79          1       3  30.000000      0      0   12.4750           1   
283         1       3  19.000000      0      0    8.0500           0   
504         1       1  16.000000      0      0   86.5000           1   
751         1       3   6.000000      0      1   12.4750           0   
147         0       3   9.000000      2      2   34.3750           1   
303         1       2  29.699118      0      0   12.3500           1   
229         0       3  29.699118      3      1   25.4667           1   
455         1       3  29.000000      0      0    7.8958           0   
219         0       2  30.000000      0      0   10.5000           0   
163         0       3  17.000000      0      0    8.6625           0   
391         1       3  21.000000      0      0    7.7958           0   
57          0       3  28.500000      0      0    7.2292           0   
139         0       1  24.000000      0      0   79.2000           0   
575         0       3  19.000000      0      0   14.5000           0   
207         1       3  26.000000      0      0   18.7875           0   
728         0       2  25.000000      1      0   26.0000           0   
..        ...     ...        ...    ...    ...       ...         ...   
755         1       2   0.670000      1      1   14.5000           0   
271         1       3  25.000000      0      0    0.0000           0   
483         1       3  63.000000      0      0    9.5875           1   
855         1       3  18.000000      0      1    9.3500           1   
600         1       2  24.000000      2      1   27.0000           1   
58          1       2   5.000000      1      2   27.7500           1   
595         0       3  36.000000      1      1   24.1500           0   
318         1       1  31.000000      0      2  164.8667           1   
121         0       3  29.699118      0      0    8.0500           0   
335         0       3  29.699118      0      0    7.8958           0   
848         0       2  28.000000      0      1   33.0000           0   
162         0       3  26.000000      0      0    7.7750           0   
760         0       3  29.699118      0      0   14.5000           0   
368         1       3  29.699118      0      0    7.7500           1   
680         0       3  29.699118      0      0    8.1375           1   
320         0       3  22.000000      0      0    7.2500           0   
170         0       1  61.000000      0      0   33.5000           0   
415         0       3  29.699118      0      0    8.0500           1   
631         0       3  51.000000      0      0    7.0542           0   
280         0       3  65.000000      0      0    7.7500           0   
636         0       3  32.000000      0      0    7.9250           0   
236         0       2  44.000000      1      0   26.0000           0   
356         1       1  22.000000      0      1   55.0000           1   
344         0       2  36.000000      0      

In [25]:
valid

Survived  Pclass        Age  SibSp  Parch      Fare  Sex_female  \
2           1       3  26.000000      0      0    7.9250           1   
3           1       1  35.000000      1      0   53.1000           1   
13          0       3  39.000000      1      5   31.2750           0   
18          0       3  31.000000      1      0   18.0000           1   
20          0       2  35.000000      0      0   26.0000           0   
24          0       3   8.000000      3      1   21.0750           1   
25          1       3  38.000000      1      5   31.3875           1   
29          0       3  29.699118      0      0    7.8958           0   
30          0       1  40.000000      0      0   27.7208           0   
36          1       3  29.699118      0      0    7.2292           0   
48          0       3  29.699118      2      0   21.6792           0   
54          0       1  65.000000      0      1   61.9792           0   
56          1       2  21.000000      0      0   10.5000           1   
59          0       3  11.000000      5      2   46.9000           0   
66          1       2  29.000000      0      0   10.5000           1   
70          0       2  32.000000      0      0   10.5000           0   
71          0       3  16.000000      5      2   46.9000           1   
73          0       3  26.000000      1      0   14.4542           0   
78          1       2   0.830000      0      2   29.0000           0   
82          1       3  29.699118      0      0    7.7875           1   
88          1       1  23.000000      3      2  263.0000           1   
89          0       3  24.000000      0      0    8.0500           0   
90          0       3  29.000000      0      0    8.0500           0   
96          0       1  71.000000      0      0   34.6542           0   
97          1       1  23.000000      0      1   63.3583           0   
99          0       2  34.000000      1      0   26.0000           0   
100         0       3  28.000000      0      0    7.8958           1   
102         0       1  21.000000      0      1   77.2875           0   
103         0       3  33.000000      0      0    8.6542           0   
109         1       3  29.699118      1      0   24.1500           1   
..        ...     ...        ...    ...    ...       ...         ...   
753         0       3  23.000000      0      0    7.8958           0   
754         1       2  48.000000      1      2   65.0000           1   
771         0       3  48.000000      0      0    7.8542           0   
780         1       3  13.000000      0      0    7.2292           1   
788         1       3   1.000000      1      2   20.5750           0   
790         0       3  29.699118      0      0    7.7500           0   
791         0       2  16.000000      0      0   26.0000           0   
793         0       1  29.699118      0      0   30.6958           0   
803         1       3   0.420000      0      1    8.5167           0   
807         0       3  18.000000      0      0    7.7750           1   
811         0       3  39.000000      0      0   24.1500           0   
816         0       3  23.000000      0      0    7.9250           1   
818         0       3  43.000000      0      0    6.4500           0   
823         1       3  27.000000      0      1   12.4750           1   
828         1       3  29.699118      0      0    7.7500           0   
835         1       1  39.000000      1      1   83.1583           1   
842         1       1  30.000000      0      0   31.0000           1   
845         0       3  42.000000      0      0    7.5500           0   
846         0       3  29.699118      8      2   69.5500           0   
847         0       3  35.000000      0      0    7.8958           0   
851         0       3  74.000000      0      0    7.7750           0   
858         1       3  24.000000      0      3   19.2583           1   
867         0       1  31.000000      0      0   50.4958           0   
868         0       3  29.699118      0      

In [26]:
train_x = train.drop(columns=['Survived'])
train_y = train[['Survived']]
valid_x = valid.drop(columns=['Survived'])
valid_y = valid[['Survived']]

In [27]:
import keras

In [28]:
from keras.models import Sequential   #input부터 차례대로 Layer를 쌓아주겠다???

In [29]:
nn_model = Sequential()

In [30]:
from keras.layers import InputLayer, Dense   #InputLayer는 말 그대로, Dnese는 Hidden layer

In [31]:
nn_model.add(InputLayer(input_shape=(10,)))

In [32]:
nn_model.summary()    #지금까지 어떤 형식의 모델이 생겼는지 볼 수 있음.

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
Total params: 0
Trainable params: 0
Non-trainable params: 0
_________________________________________________________________


In [33]:
nn_model.add(Dense(units=10, activation = 'sigmoid'))

- units: unit의 개수
- activation: activation 함수<br>
Dense를 통해 hidden layer를 쌓을 수 있음.

In [34]:
nn_model.add(Dense(units=6, activation = 'sigmoid'))   #layer의 수 점점 쌓으면 됨

In [35]:
nn_model.add(Dense(units=3, activation = 'sigmoid'))

In [36]:
nn_model.add(Dense(units=1, activation = 'sigmoid'))

In [37]:
nn_model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 10)                110       
_________________________________________________________________
dense_2 (Dense)              (None, 6)                 66        
_________________________________________________________________
dense_3 (Dense)              (None, 3)                 21        
_________________________________________________________________
dense_4 (Dense)              (None, 1)                 4         
Total params: 201
Trainable params: 201
Non-trainable params: 0
_________________________________________________________________


summary에서 Param#: weight의 개수. input층의 weight이 110개인 이유는 10개의 bias뭐시기가 있기 때문.

In [38]:
nn_model.compile(optimizer='sgd', loss='mean_squared_error', metrics=['accuracy'])

optimizer: SGD(stochastic gradient descent)<br>
loss: 에러를 어떻게 정할 것인가<br>
metrics: 어떻게 평가할 것인가. 평가척도

In [40]:
nn_model.fit(x=train_x, y=train_y, epochs=1000, 
             validation_data=(valid_x, valid_y))  #epochs: 몇 번 학습시킬 것인가

Train on 713 samples, validate on 178 samples
Epoch 1/1000
713/713 [==============================] - 1s 2ms/step - loss: 0.2420 - acc: 0.6087 - val_loss: 0.2290 - val_acc: 0.6461
Epoch 2/1000
713/713 [==============================] - 0s 139us/step - loss: 0.2416 - acc: 0.6087 - val_loss: 0.2289 - val_acc: 0.6461
Epoch 3/1000
713/713 [==============================] - 0s 142us/step - loss: 0.2413 - acc: 0.6087 - val_loss: 0.2288 - val_acc: 0.6461
Epoch 4/1000
713/713 [==============================] - 0s 152us/step - loss: 0.2410 - acc: 0.6087 - val_loss: 0.2287 - val_acc: 0.6461
Epoch 5/1000
713/713 [==============================] - 0s 147us/step - loss: 0.2408 - acc: 0.6087 - val_loss: 0.2286 - val_acc: 0.6461
Epoch 6/1000
713/713 [==============================] - 0s 160us/step - loss: 0.2405 - acc: 0.6087 - val_loss: 0.2285 - val_acc: 0.6461
Epoch 7/1000
713/713 [==============================] - 0s 152us/step - loss: 0.2403 - acc: 0.6087 - val_loss: 0.2284 - val_acc: 0.6461
Epoc

713/713 [==============================] - 0s 111us/step - loss: 0.2376 - acc: 0.6087 - val_loss: 0.2292 - val_acc: 0.6461
Epoch 61/1000
713/713 [==============================] - 0s 98us/step - loss: 0.2375 - acc: 0.6087 - val_loss: 0.2292 - val_acc: 0.6461
Epoch 62/1000
713/713 [==============================] - 0s 98us/step - loss: 0.2375 - acc: 0.6087 - val_loss: 0.2291 - val_acc: 0.6461
Epoch 63/1000
713/713 [==============================] - 0s 84us/step - loss: 0.2375 - acc: 0.6087 - val_loss: 0.2292 - val_acc: 0.6461
Epoch 64/1000
713/713 [==============================] - 0s 86us/step - loss: 0.2375 - acc: 0.6087 - val_loss: 0.2292 - val_acc: 0.6461
Epoch 65/1000
713/713 [==============================] - 0s 107us/step - loss: 0.2375 - acc: 0.6087 - val_loss: 0.2292 - val_acc: 0.6461
Epoch 66/1000
713/713 [==============================] - 0s 112us/step - loss: 0.2375 - acc: 0.6087 - val_loss: 0.2292 - val_acc: 0.6461
Epoch 67/1000
713/713 [==============================] - 0s

713/713 [==============================] - 0s 77us/step - loss: 0.2372 - acc: 0.6087 - val_loss: 0.2290 - val_acc: 0.6461
Epoch 121/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2372 - acc: 0.6087 - val_loss: 0.2290 - val_acc: 0.6461
Epoch 122/1000
713/713 [==============================] - 0s 72us/step - loss: 0.2372 - acc: 0.6087 - val_loss: 0.2290 - val_acc: 0.6461
Epoch 123/1000
713/713 [==============================] - 0s 66us/step - loss: 0.2372 - acc: 0.6087 - val_loss: 0.2290 - val_acc: 0.6461
Epoch 124/1000
713/713 [==============================] - 0s 83us/step - loss: 0.2372 - acc: 0.6087 - val_loss: 0.2290 - val_acc: 0.6461
Epoch 125/1000
713/713 [==============================] - 0s 77us/step - loss: 0.2372 - acc: 0.6087 - val_loss: 0.2290 - val_acc: 0.6461
Epoch 126/1000
713/713 [==============================] - 0s 76us/step - loss: 0.2372 - acc: 0.6087 - val_loss: 0.2290 - val_acc: 0.6461
Epoch 127/1000
713/713 [==============================] 

Epoch 180/1000
713/713 [==============================] - 0s 59us/step - loss: 0.2368 - acc: 0.6087 - val_loss: 0.2287 - val_acc: 0.6461
Epoch 181/1000
713/713 [==============================] - 0s 65us/step - loss: 0.2368 - acc: 0.6087 - val_loss: 0.2287 - val_acc: 0.6461
Epoch 182/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2368 - acc: 0.6087 - val_loss: 0.2287 - val_acc: 0.6461
Epoch 183/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2368 - acc: 0.6087 - val_loss: 0.2287 - val_acc: 0.6461
Epoch 184/1000
713/713 [==============================] - 0s 63us/step - loss: 0.2368 - acc: 0.6087 - val_loss: 0.2287 - val_acc: 0.6461
Epoch 185/1000
713/713 [==============================] - 0s 66us/step - loss: 0.2368 - acc: 0.6087 - val_loss: 0.2286 - val_acc: 0.6461
Epoch 186/1000
713/713 [==============================] - 0s 63us/step - loss: 0.2368 - acc: 0.6087 - val_loss: 0.2287 - val_acc: 0.6461
Epoch 187/1000
713/713 [=================

Epoch 240/1000
713/713 [==============================] - 0s 65us/step - loss: 0.2364 - acc: 0.6087 - val_loss: 0.2283 - val_acc: 0.6461
Epoch 241/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2364 - acc: 0.6087 - val_loss: 0.2283 - val_acc: 0.6461
Epoch 242/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2364 - acc: 0.6087 - val_loss: 0.2284 - val_acc: 0.6461
Epoch 243/1000
713/713 [==============================] - 0s 69us/step - loss: 0.2364 - acc: 0.6087 - val_loss: 0.2284 - val_acc: 0.6461
Epoch 244/1000
713/713 [==============================] - 0s 70us/step - loss: 0.2364 - acc: 0.6087 - val_loss: 0.2283 - val_acc: 0.6461
Epoch 245/1000
713/713 [==============================] - 0s 72us/step - loss: 0.2364 - acc: 0.6087 - val_loss: 0.2283 - val_acc: 0.6461
Epoch 246/1000
713/713 [==============================] - 0s 66us/step - loss: 0.2364 - acc: 0.6087 - val_loss: 0.2283 - val_acc: 0.6461
Epoch 247/1000
713/713 [=================

Epoch 300/1000
713/713 [==============================] - 0s 56us/step - loss: 0.2359 - acc: 0.6087 - val_loss: 0.2279 - val_acc: 0.6461
Epoch 301/1000
713/713 [==============================] - 0s 59us/step - loss: 0.2359 - acc: 0.6087 - val_loss: 0.2278 - val_acc: 0.6461
Epoch 302/1000
713/713 [==============================] - 0s 56us/step - loss: 0.2359 - acc: 0.6087 - val_loss: 0.2278 - val_acc: 0.6461
Epoch 303/1000
713/713 [==============================] - 0s 69us/step - loss: 0.2359 - acc: 0.6087 - val_loss: 0.2278 - val_acc: 0.6461
Epoch 304/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2359 - acc: 0.6087 - val_loss: 0.2279 - val_acc: 0.6461
Epoch 305/1000
713/713 [==============================] - 0s 63us/step - loss: 0.2359 - acc: 0.6087 - val_loss: 0.2279 - val_acc: 0.6461
Epoch 306/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2359 - acc: 0.6087 - val_loss: 0.2279 - val_acc: 0.6461
Epoch 307/1000
713/713 [=================

Epoch 360/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2354 - acc: 0.6087 - val_loss: 0.2273 - val_acc: 0.6461
Epoch 361/1000
713/713 [==============================] - 0s 81us/step - loss: 0.2354 - acc: 0.6087 - val_loss: 0.2273 - val_acc: 0.6461
Epoch 362/1000
713/713 [==============================] - 0s 74us/step - loss: 0.2354 - acc: 0.6087 - val_loss: 0.2272 - val_acc: 0.6461
Epoch 363/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2354 - acc: 0.6087 - val_loss: 0.2272 - val_acc: 0.6461
Epoch 364/1000
713/713 [==============================] - 0s 69us/step - loss: 0.2354 - acc: 0.6087 - val_loss: 0.2272 - val_acc: 0.6461
Epoch 365/1000
713/713 [==============================] - 0s 72us/step - loss: 0.2354 - acc: 0.6087 - val_loss: 0.2273 - val_acc: 0.6461
Epoch 366/1000
713/713 [==============================] - 0s 70us/step - loss: 0.2353 - acc: 0.6087 - val_loss: 0.2272 - val_acc: 0.6461
Epoch 367/1000
713/713 [=================

Epoch 420/1000
713/713 [==============================] - 0s 59us/step - loss: 0.2348 - acc: 0.6087 - val_loss: 0.2266 - val_acc: 0.6461
Epoch 421/1000
713/713 [==============================] - 0s 59us/step - loss: 0.2347 - acc: 0.6087 - val_loss: 0.2265 - val_acc: 0.6461
Epoch 422/1000
713/713 [==============================] - 0s 74us/step - loss: 0.2347 - acc: 0.6087 - val_loss: 0.2265 - val_acc: 0.6461
Epoch 423/1000
713/713 [==============================] - 0s 72us/step - loss: 0.2347 - acc: 0.6087 - val_loss: 0.2265 - val_acc: 0.6461
Epoch 424/1000
713/713 [==============================] - 0s 72us/step - loss: 0.2347 - acc: 0.6087 - val_loss: 0.2265 - val_acc: 0.6461
Epoch 425/1000
713/713 [==============================] - 0s 72us/step - loss: 0.2347 - acc: 0.6087 - val_loss: 0.2265 - val_acc: 0.6461
Epoch 426/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2347 - acc: 0.6087 - val_loss: 0.2264 - val_acc: 0.6461
Epoch 427/1000
713/713 [=================

Epoch 480/1000
713/713 [==============================] - 0s 56us/step - loss: 0.2340 - acc: 0.6087 - val_loss: 0.2258 - val_acc: 0.6461
Epoch 481/1000
713/713 [==============================] - 0s 62us/step - loss: 0.2340 - acc: 0.6087 - val_loss: 0.2258 - val_acc: 0.6461
Epoch 482/1000
713/713 [==============================] - 0s 52us/step - loss: 0.2340 - acc: 0.6087 - val_loss: 0.2258 - val_acc: 0.6461
Epoch 483/1000
713/713 [==============================] - 0s 65us/step - loss: 0.2340 - acc: 0.6087 - val_loss: 0.2257 - val_acc: 0.6461
Epoch 484/1000
713/713 [==============================] - 0s 62us/step - loss: 0.2340 - acc: 0.6087 - val_loss: 0.2257 - val_acc: 0.6461
Epoch 485/1000
713/713 [==============================] - 0s 77us/step - loss: 0.2340 - acc: 0.6087 - val_loss: 0.2257 - val_acc: 0.6461
Epoch 486/1000
713/713 [==============================] - 0s 70us/step - loss: 0.2340 - acc: 0.6087 - val_loss: 0.2257 - val_acc: 0.6461
Epoch 487/1000
713/713 [=================

Epoch 540/1000
713/713 [==============================] - 0s 58us/step - loss: 0.2333 - acc: 0.6087 - val_loss: 0.2248 - val_acc: 0.6461
Epoch 541/1000
713/713 [==============================] - 0s 62us/step - loss: 0.2333 - acc: 0.6087 - val_loss: 0.2248 - val_acc: 0.6461
Epoch 542/1000
713/713 [==============================] - 0s 53us/step - loss: 0.2332 - acc: 0.6087 - val_loss: 0.2248 - val_acc: 0.6461
Epoch 543/1000
713/713 [==============================] - 0s 73us/step - loss: 0.2332 - acc: 0.6087 - val_loss: 0.2248 - val_acc: 0.6461
Epoch 544/1000
713/713 [==============================] - 0s 65us/step - loss: 0.2332 - acc: 0.6087 - val_loss: 0.2247 - val_acc: 0.6461
Epoch 545/1000
713/713 [==============================] - 0s 73us/step - loss: 0.2332 - acc: 0.6087 - val_loss: 0.2247 - val_acc: 0.6461
Epoch 546/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2332 - acc: 0.6087 - val_loss: 0.2247 - val_acc: 0.6461
Epoch 547/1000
713/713 [=================

Epoch 600/1000
713/713 [==============================] - 0s 58us/step - loss: 0.2324 - acc: 0.6087 - val_loss: 0.2238 - val_acc: 0.6461
Epoch 601/1000
713/713 [==============================] - 0s 55us/step - loss: 0.2324 - acc: 0.6087 - val_loss: 0.2238 - val_acc: 0.6461
Epoch 602/1000
713/713 [==============================] - 0s 51us/step - loss: 0.2323 - acc: 0.6087 - val_loss: 0.2238 - val_acc: 0.6461
Epoch 603/1000
713/713 [==============================] - 0s 59us/step - loss: 0.2323 - acc: 0.6087 - val_loss: 0.2237 - val_acc: 0.6461
Epoch 604/1000
713/713 [==============================] - 0s 65us/step - loss: 0.2323 - acc: 0.6087 - val_loss: 0.2237 - val_acc: 0.6461
Epoch 605/1000
713/713 [==============================] - 0s 66us/step - loss: 0.2323 - acc: 0.6087 - val_loss: 0.2237 - val_acc: 0.6461
Epoch 606/1000
713/713 [==============================] - 0s 65us/step - loss: 0.2323 - acc: 0.6087 - val_loss: 0.2237 - val_acc: 0.6461
Epoch 607/1000
713/713 [=================

Epoch 660/1000
713/713 [==============================] - 0s 53us/step - loss: 0.2313 - acc: 0.6087 - val_loss: 0.2226 - val_acc: 0.6461
Epoch 661/1000
713/713 [==============================] - 0s 55us/step - loss: 0.2313 - acc: 0.6087 - val_loss: 0.2226 - val_acc: 0.6461
Epoch 662/1000
713/713 [==============================] - 0s 48us/step - loss: 0.2313 - acc: 0.6087 - val_loss: 0.2226 - val_acc: 0.6461
Epoch 663/1000
713/713 [==============================] - 0s 62us/step - loss: 0.2313 - acc: 0.6087 - val_loss: 0.2226 - val_acc: 0.6461
Epoch 664/1000
713/713 [==============================] - 0s 65us/step - loss: 0.2313 - acc: 0.6087 - val_loss: 0.2226 - val_acc: 0.6461
Epoch 665/1000
713/713 [==============================] - 0s 69us/step - loss: 0.2312 - acc: 0.6087 - val_loss: 0.2226 - val_acc: 0.6461
Epoch 666/1000
713/713 [==============================] - 0s 63us/step - loss: 0.2312 - acc: 0.6087 - val_loss: 0.2226 - val_acc: 0.6461
Epoch 667/1000
713/713 [=================

Epoch 720/1000
713/713 [==============================] - 0s 55us/step - loss: 0.2301 - acc: 0.6087 - val_loss: 0.2214 - val_acc: 0.6461
Epoch 721/1000
713/713 [==============================] - 0s 48us/step - loss: 0.2301 - acc: 0.6087 - val_loss: 0.2214 - val_acc: 0.6461
Epoch 722/1000
713/713 [==============================] - 0s 63us/step - loss: 0.2301 - acc: 0.6087 - val_loss: 0.2213 - val_acc: 0.6461
Epoch 723/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2301 - acc: 0.6087 - val_loss: 0.2213 - val_acc: 0.6461
Epoch 724/1000
713/713 [==============================] - 0s 62us/step - loss: 0.2301 - acc: 0.6087 - val_loss: 0.2213 - val_acc: 0.6461
Epoch 725/1000
713/713 [==============================] - 0s 62us/step - loss: 0.2300 - acc: 0.6087 - val_loss: 0.2213 - val_acc: 0.6461
Epoch 726/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2300 - acc: 0.6087 - val_loss: 0.2212 - val_acc: 0.6461
Epoch 727/1000
713/713 [=================

Epoch 780/1000
713/713 [==============================] - 0s 69us/step - loss: 0.2288 - acc: 0.6087 - val_loss: 0.2199 - val_acc: 0.6461
Epoch 781/1000
713/713 [==============================] - 0s 52us/step - loss: 0.2288 - acc: 0.6087 - val_loss: 0.2199 - val_acc: 0.6461
Epoch 782/1000
713/713 [==============================] - 0s 59us/step - loss: 0.2288 - acc: 0.6087 - val_loss: 0.2198 - val_acc: 0.6461
Epoch 783/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2288 - acc: 0.6087 - val_loss: 0.2198 - val_acc: 0.6461
Epoch 784/1000
713/713 [==============================] - 0s 58us/step - loss: 0.2287 - acc: 0.6087 - val_loss: 0.2198 - val_acc: 0.6461
Epoch 785/1000
713/713 [==============================] - 0s 56us/step - loss: 0.2287 - acc: 0.6087 - val_loss: 0.2198 - val_acc: 0.6461
Epoch 786/1000
713/713 [==============================] - 0s 58us/step - loss: 0.2287 - acc: 0.6087 - val_loss: 0.2198 - val_acc: 0.6461
Epoch 787/1000
713/713 [=================

Epoch 840/1000
713/713 [==============================] - 0s 53us/step - loss: 0.2273 - acc: 0.6087 - val_loss: 0.2181 - val_acc: 0.6461
Epoch 841/1000
713/713 [==============================] - 0s 49us/step - loss: 0.2273 - acc: 0.6087 - val_loss: 0.2180 - val_acc: 0.6461
Epoch 842/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2273 - acc: 0.6087 - val_loss: 0.2180 - val_acc: 0.6461
Epoch 843/1000
713/713 [==============================] - 0s 65us/step - loss: 0.2273 - acc: 0.6087 - val_loss: 0.2180 - val_acc: 0.6461
Epoch 844/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2272 - acc: 0.6087 - val_loss: 0.2180 - val_acc: 0.6461
Epoch 845/1000
713/713 [==============================] - 0s 56us/step - loss: 0.2272 - acc: 0.6087 - val_loss: 0.2180 - val_acc: 0.6461
Epoch 846/1000
713/713 [==============================] - 0s 65us/step - loss: 0.2272 - acc: 0.6087 - val_loss: 0.2180 - val_acc: 0.6461
Epoch 847/1000
713/713 [=================

Epoch 900/1000
713/713 [==============================] - 0s 56us/step - loss: 0.2257 - acc: 0.6087 - val_loss: 0.2163 - val_acc: 0.6461
Epoch 901/1000
713/713 [==============================] - 0s 58us/step - loss: 0.2257 - acc: 0.6087 - val_loss: 0.2163 - val_acc: 0.6461
Epoch 902/1000
713/713 [==============================] - 0s 62us/step - loss: 0.2257 - acc: 0.6087 - val_loss: 0.2163 - val_acc: 0.6461
Epoch 903/1000
713/713 [==============================] - 0s 66us/step - loss: 0.2257 - acc: 0.6087 - val_loss: 0.2162 - val_acc: 0.6461
Epoch 904/1000
713/713 [==============================] - 0s 62us/step - loss: 0.2256 - acc: 0.6087 - val_loss: 0.2162 - val_acc: 0.6461
Epoch 905/1000
713/713 [==============================] - 0s 59us/step - loss: 0.2256 - acc: 0.6087 - val_loss: 0.2161 - val_acc: 0.6461
Epoch 906/1000
713/713 [==============================] - 0s 63us/step - loss: 0.2256 - acc: 0.6087 - val_loss: 0.2161 - val_acc: 0.6461
Epoch 907/1000
713/713 [=================

Epoch 960/1000
713/713 [==============================] - 0s 70us/step - loss: 0.2241 - acc: 0.6087 - val_loss: 0.2144 - val_acc: 0.6461
Epoch 961/1000
713/713 [==============================] - 0s 52us/step - loss: 0.2240 - acc: 0.6087 - val_loss: 0.2143 - val_acc: 0.6461
Epoch 962/1000
713/713 [==============================] - 0s 55us/step - loss: 0.2240 - acc: 0.6087 - val_loss: 0.2143 - val_acc: 0.6461
Epoch 963/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2240 - acc: 0.6087 - val_loss: 0.2143 - val_acc: 0.6461
Epoch 964/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2240 - acc: 0.6087 - val_loss: 0.2143 - val_acc: 0.6461
Epoch 965/1000
713/713 [==============================] - 0s 53us/step - loss: 0.2239 - acc: 0.6087 - val_loss: 0.2142 - val_acc: 0.6461
Epoch 966/1000
713/713 [==============================] - 0s 53us/step - loss: 0.2239 - acc: 0.6087 - val_loss: 0.2142 - val_acc: 0.6461
Epoch 967/1000
713/713 [=================

이렇게 돌려보면 accuracy가 별로 좋지 않음. 매우 전통적인 모델이라...

***
### Improve models
1. Activation Function: Single perceptron에 나타난 결과를 어떻게 변형시킬 것인가.
    - Binary activation: 특정 역치값을 넘으면 활성화, 안넘으면 비활성화. 그러나 수학적인 관점에서 보면 이러한 activation함수를 가지고 활성화를 하는 것에 문제가 있음. 미분값이 항상 0이기 때문에 고정된 값을 이용해 update를 수행해야 함.
    - Sigmoid activation: 0과 1 사이의 실수 값을 가짐. 미분이 굉장히 쉬우며 binary activation과 유사한 역할을 함.
    - Tanh(탄젠트) Activation: Sigmoid와 유사하지만 -1과 1 사이의 값을 가짐. 미분이 간단하고 sigmoid보다 미분값이 더 크기 때문에 weight값이 더 빠르게 수정됨.<br>
-> Vanishing gradient problem: layer가 깊어질 수록 gradient값이 점점 작아짐. output layer에서 멀어질수록 weight값이 거의 업데이트되지 않음. sigmoid를 여러 개 겹치다보면 미분값이 점점 0으로 수렴하게 됨.
    - ReLU activation: 0보다 클 때는 y=ax에 따르고, 0보다 작을 때는 무조건 0. 즉, 0보다 작으면 training을 하지 않고, 0보다 크면 고정된 값(a)으로 training을 함. 그러나 0일 때 미분이 불가능하다는 단점이 있으며, 초기 weight값 지정에 많은 영향을 받음.
<br>
<br>
2. Overfit and Underfit: 두 경우에 빠지지 않았을 경우를 general한 경우라고 일컬음.
    - SVM: 두 class 사이에 선을 그을 때, 각 class와의 간격(gap)이 최대가 되도록 선을 긋는 방법. deep learning 이전에 generalization 관점에서 매우 유용했던 모델.
    - Regularization: 모델에 일부러 패널티를 부여. error를 계산할 때 x들의 기울기에 따른 패널티를 줘서 모델을 좀 더 general하게 만듦.
    - Dropout: 랜덤하게 일부 node를 제외하고 학습시키는 모델. 
        - Co-adaption(지나치게 모델에 적합화되는 것)회피: 영향력이 강한 노드를 제외시켜서 다른 노드도 학습할 수 있는 기회를 줌.
        - Ensemble효과(여러 개의 모델을 합쳐서 하나로 만드는 것): 각 batch에 따라 학습된 여러 모델을 합쳐서 하나의 모델로 만듦. 각 batch는 서로 다른 dropout을 진행한 모델들.
<br>
<br>
3. Optimizer
    - SGD: Stochastic Gradient Descent(경사 하강법). 매 epcoh(학습)마다 error값을 계산하고, error의 미분값에 따라 error값이 점점 작아지는 형태로 모델을 수정하는 기법. minimum값이 될 때까지 진행.<br>
    -> 지역 최소값(local minima)에 빠지는 문제점이 발생.
        - momentum optimizer: 가속도값을 이용해서 원래 이동하려는 방향으로 계속 나아가려는 성질을 통해, 얕은 local minima를 빠져나갈 수 있도록 학습.
        - Various optimizer: Adagrad, Rmsprop, Adam 등.
        - batch training: training할 때 데이터를 하나씩 넣어서 학습시킬 수 있고, 전체 데이터를 넣어서 학습시킬 수 있음. batch training은 데이터의 일부분만 떼어서 training을 하는 방법. local minima에 빠질 가능성이 적음.

In [41]:
nn_model2 = Sequential()

In [42]:
nn_model2.add(InputLayer((10,)))

In [43]:
nn_model2.add(Dense(units=10, activation='relu'))

In [44]:
nn_model2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_5 (Dense)              (None, 10)                110       
Total params: 110
Trainable params: 110
Non-trainable params: 0
_________________________________________________________________


In [45]:
nn_model2.add(Dense(units=6, activation='relu'))
nn_model2.add(Dense(units=3, activation='relu'))

In [46]:
nn_model2.add(Dense(units=1, activation='sigmoid'))

마지막에는 relu가 아니라 sigmoid를 사용해줌. relu는 0과 1사이의 값이 아니라 발산하는 형태의 함수이기 때문에.

In [47]:
nn_model2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_5 (Dense)              (None, 10)                110       
_________________________________________________________________
dense_6 (Dense)              (None, 6)                 66        
_________________________________________________________________
dense_7 (Dense)              (None, 3)                 21        
_________________________________________________________________
dense_8 (Dense)              (None, 1)                 4         
Total params: 201
Trainable params: 201
Non-trainable params: 0
_________________________________________________________________


In [50]:
nn_model2.compile(optimizer='sgd', loss='mean_squared_error', metrics=['accuracy'])

In [51]:
nn_model2.fit(x=train_x, y=train_y, epochs=1000,
             validation_data=(valid_x, valid_y))    #학습 실행

Train on 713 samples, validate on 178 samples
Epoch 1/1000
713/713 [==============================] - 1s 2ms/step - loss: 0.2735 - acc: 0.3913 - val_loss: 0.2642 - val_acc: 0.3539
Epoch 2/1000
713/713 [==============================] - 0s 133us/step - loss: 0.2549 - acc: 0.4558 - val_loss: 0.2537 - val_acc: 0.6517
Epoch 3/1000
713/713 [==============================] - 0s 132us/step - loss: 0.2486 - acc: 0.6732 - val_loss: 0.2499 - val_acc: 0.6742
Epoch 4/1000
713/713 [==============================] - 0s 126us/step - loss: 0.2464 - acc: 0.6648 - val_loss: 0.2483 - val_acc: 0.6742
Epoch 5/1000
713/713 [==============================] - 0s 140us/step - loss: 0.2450 - acc: 0.6690 - val_loss: 0.2472 - val_acc: 0.6742
Epoch 6/1000
713/713 [==============================] - 0s 131us/step - loss: 0.2434 - acc: 0.6662 - val_loss: 0.2459 - val_acc: 0.6742
Epoch 7/1000
713/713 [==============================] - 0s 122us/step - loss: 0.2420 - acc: 0.6732 - val_loss: 0.2443 - val_acc: 0.6798
Epoc

Epoch 61/1000
713/713 [==============================] - 0s 80us/step - loss: 0.2190 - acc: 0.6690 - val_loss: 0.2174 - val_acc: 0.6798
Epoch 62/1000
713/713 [==============================] - 0s 83us/step - loss: 0.2189 - acc: 0.6732 - val_loss: 0.2171 - val_acc: 0.6798
Epoch 63/1000
713/713 [==============================] - 0s 95us/step - loss: 0.2188 - acc: 0.6704 - val_loss: 0.2169 - val_acc: 0.6798
Epoch 64/1000
713/713 [==============================] - 0s 112us/step - loss: 0.2186 - acc: 0.6788 - val_loss: 0.2168 - val_acc: 0.6742
Epoch 65/1000
713/713 [==============================] - 0s 81us/step - loss: 0.2185 - acc: 0.6676 - val_loss: 0.2166 - val_acc: 0.6685
Epoch 66/1000
713/713 [==============================] - 0s 77us/step - loss: 0.2183 - acc: 0.6662 - val_loss: 0.2164 - val_acc: 0.6742
Epoch 67/1000
713/713 [==============================] - 0s 108us/step - loss: 0.2182 - acc: 0.6690 - val_loss: 0.2162 - val_acc: 0.6685
Epoch 68/1000
713/713 [=======================

713/713 [==============================] - 0s 69us/step - loss: 0.2148 - acc: 0.6718 - val_loss: 0.2108 - val_acc: 0.6798
Epoch 122/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2148 - acc: 0.6732 - val_loss: 0.2106 - val_acc: 0.6742
Epoch 123/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2147 - acc: 0.6746 - val_loss: 0.2106 - val_acc: 0.6798
Epoch 124/1000
713/713 [==============================] - 0s 67us/step - loss: 0.2146 - acc: 0.6774 - val_loss: 0.2106 - val_acc: 0.6798
Epoch 125/1000
713/713 [==============================] - 0s 69us/step - loss: 0.2146 - acc: 0.6746 - val_loss: 0.2103 - val_acc: 0.6742
Epoch 126/1000
713/713 [==============================] - 0s 76us/step - loss: 0.2145 - acc: 0.6788 - val_loss: 0.2102 - val_acc: 0.6742
Epoch 127/1000
713/713 [==============================] - 0s 73us/step - loss: 0.2144 - acc: 0.6774 - val_loss: 0.2102 - val_acc: 0.6742
Epoch 128/1000
713/713 [==============================] 

Epoch 181/1000
713/713 [==============================] - 0s 56us/step - loss: 0.2121 - acc: 0.6774 - val_loss: 0.2055 - val_acc: 0.6798
Epoch 182/1000
713/713 [==============================] - 0s 51us/step - loss: 0.2118 - acc: 0.6788 - val_loss: 0.2056 - val_acc: 0.6854
Epoch 183/1000
713/713 [==============================] - 0s 63us/step - loss: 0.2123 - acc: 0.6746 - val_loss: 0.2054 - val_acc: 0.6798
Epoch 184/1000
713/713 [==============================] - 0s 74us/step - loss: 0.2117 - acc: 0.6802 - val_loss: 0.2056 - val_acc: 0.6798
Epoch 185/1000
713/713 [==============================] - 0s 77us/step - loss: 0.2119 - acc: 0.6788 - val_loss: 0.2052 - val_acc: 0.6742
Epoch 186/1000
713/713 [==============================] - 0s 77us/step - loss: 0.2115 - acc: 0.6830 - val_loss: 0.2055 - val_acc: 0.6854
Epoch 187/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2119 - acc: 0.6802 - val_loss: 0.2048 - val_acc: 0.6854
Epoch 188/1000
713/713 [=================

Epoch 241/1000
713/713 [==============================] - 0s 56us/step - loss: 0.2063 - acc: 0.6872 - val_loss: 0.1940 - val_acc: 0.7079
Epoch 242/1000
713/713 [==============================] - 0s 52us/step - loss: 0.2054 - acc: 0.6858 - val_loss: 0.1952 - val_acc: 0.7022
Epoch 243/1000
713/713 [==============================] - 0s 60us/step - loss: 0.2057 - acc: 0.6774 - val_loss: 0.1974 - val_acc: 0.6910
Epoch 244/1000
713/713 [==============================] - 0s 58us/step - loss: 0.2059 - acc: 0.6816 - val_loss: 0.1938 - val_acc: 0.7079
Epoch 245/1000
713/713 [==============================] - 0s 59us/step - loss: 0.2047 - acc: 0.6718 - val_loss: 0.1934 - val_acc: 0.7135
Epoch 246/1000
713/713 [==============================] - 0s 59us/step - loss: 0.2048 - acc: 0.6858 - val_loss: 0.1926 - val_acc: 0.7135
Epoch 247/1000
713/713 [==============================] - 0s 72us/step - loss: 0.2047 - acc: 0.6760 - val_loss: 0.1933 - val_acc: 0.7022
Epoch 248/1000
713/713 [=================

Epoch 301/1000
713/713 [==============================] - 0s 58us/step - loss: 0.1879 - acc: 0.7307 - val_loss: 0.2359 - val_acc: 0.6573
Epoch 302/1000
713/713 [==============================] - 0s 51us/step - loss: 0.1926 - acc: 0.7195 - val_loss: 0.1732 - val_acc: 0.7416
Epoch 303/1000
713/713 [==============================] - 0s 62us/step - loss: 0.1919 - acc: 0.7139 - val_loss: 0.1812 - val_acc: 0.7191
Epoch 304/1000
713/713 [==============================] - 0s 65us/step - loss: 0.1903 - acc: 0.7195 - val_loss: 0.1709 - val_acc: 0.7528
Epoch 305/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1869 - acc: 0.7405 - val_loss: 0.1743 - val_acc: 0.7697
Epoch 306/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1913 - acc: 0.7125 - val_loss: 0.2017 - val_acc: 0.6517
Epoch 307/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1954 - acc: 0.7027 - val_loss: 0.2295 - val_acc: 0.6798
Epoch 308/1000
713/713 [=================

713/713 [==============================] - 0s 93us/step - loss: 0.1838 - acc: 0.7405 - val_loss: 0.1526 - val_acc: 0.8202
Epoch 361/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1730 - acc: 0.7616 - val_loss: 0.1523 - val_acc: 0.8090
Epoch 362/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1755 - acc: 0.7574 - val_loss: 0.1656 - val_acc: 0.7809
Epoch 363/1000
713/713 [==============================] - 0s 53us/step - loss: 0.1816 - acc: 0.7433 - val_loss: 0.2943 - val_acc: 0.5955
Epoch 364/1000
713/713 [==============================] - 0s 62us/step - loss: 0.1818 - acc: 0.7405 - val_loss: 0.1963 - val_acc: 0.6854
Epoch 365/1000
713/713 [==============================] - 0s 66us/step - loss: 0.1744 - acc: 0.7728 - val_loss: 0.1514 - val_acc: 0.8258
Epoch 366/1000
713/713 [==============================] - 0s 65us/step - loss: 0.1737 - acc: 0.7644 - val_loss: 0.2536 - val_acc: 0.6517
Epoch 367/1000
713/713 [==============================] 

Epoch 420/1000
713/713 [==============================] - 0s 53us/step - loss: 0.1745 - acc: 0.7588 - val_loss: 0.1519 - val_acc: 0.8034
Epoch 421/1000
713/713 [==============================] - 0s 58us/step - loss: 0.1739 - acc: 0.7644 - val_loss: 0.1444 - val_acc: 0.8258
Epoch 422/1000
713/713 [==============================] - 0s 49us/step - loss: 0.1712 - acc: 0.7700 - val_loss: 0.2077 - val_acc: 0.6685
Epoch 423/1000
713/713 [==============================] - 0s 63us/step - loss: 0.1723 - acc: 0.7742 - val_loss: 0.1986 - val_acc: 0.7079
Epoch 424/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1704 - acc: 0.7616 - val_loss: 0.1784 - val_acc: 0.7809
Epoch 425/1000
713/713 [==============================] - 0s 63us/step - loss: 0.1787 - acc: 0.7504 - val_loss: 0.1440 - val_acc: 0.8090
Epoch 426/1000
713/713 [==============================] - 0s 65us/step - loss: 0.1740 - acc: 0.7546 - val_loss: 0.1901 - val_acc: 0.6966
Epoch 427/1000
713/713 [=================

Epoch 480/1000
713/713 [==============================] - 0s 56us/step - loss: 0.1634 - acc: 0.7756 - val_loss: 0.1436 - val_acc: 0.8258
Epoch 481/1000
713/713 [==============================] - 0s 62us/step - loss: 0.1672 - acc: 0.7728 - val_loss: 0.1385 - val_acc: 0.8427
Epoch 482/1000
713/713 [==============================] - 0s 51us/step - loss: 0.1693 - acc: 0.7686 - val_loss: 0.1385 - val_acc: 0.8315
Epoch 483/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1653 - acc: 0.7812 - val_loss: 0.1483 - val_acc: 0.7921
Epoch 484/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1664 - acc: 0.7714 - val_loss: 0.1505 - val_acc: 0.7697
Epoch 485/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1684 - acc: 0.7644 - val_loss: 0.1649 - val_acc: 0.7809
Epoch 486/1000
713/713 [==============================] - 0s 59us/step - loss: 0.1640 - acc: 0.7728 - val_loss: 0.1436 - val_acc: 0.8146
Epoch 487/1000
713/713 [=================

Epoch 540/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1674 - acc: 0.7630 - val_loss: 0.1529 - val_acc: 0.7921
Epoch 541/1000
713/713 [==============================] - 0s 52us/step - loss: 0.1657 - acc: 0.7728 - val_loss: 0.1691 - val_acc: 0.7303
Epoch 542/1000
713/713 [==============================] - 0s 66us/step - loss: 0.1622 - acc: 0.7826 - val_loss: 0.1377 - val_acc: 0.8146
Epoch 543/1000
713/713 [==============================] - 0s 63us/step - loss: 0.1670 - acc: 0.7742 - val_loss: 0.1643 - val_acc: 0.7753
Epoch 544/1000
713/713 [==============================] - 0s 56us/step - loss: 0.1635 - acc: 0.7784 - val_loss: 0.1936 - val_acc: 0.7753
Epoch 545/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1706 - acc: 0.7602 - val_loss: 0.1446 - val_acc: 0.8090
Epoch 546/1000
713/713 [==============================] - 0s 62us/step - loss: 0.1668 - acc: 0.7644 - val_loss: 0.1516 - val_acc: 0.7978
Epoch 547/1000
713/713 [=================

Epoch 600/1000
713/713 [==============================] - 0s 49us/step - loss: 0.1671 - acc: 0.7672 - val_loss: 0.1389 - val_acc: 0.8146
Epoch 601/1000
713/713 [==============================] - 0s 53us/step - loss: 0.1612 - acc: 0.7756 - val_loss: 0.1341 - val_acc: 0.8258
Epoch 602/1000
713/713 [==============================] - 0s 53us/step - loss: 0.1610 - acc: 0.7840 - val_loss: 0.3159 - val_acc: 0.5281
Epoch 603/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1658 - acc: 0.7686 - val_loss: 0.1615 - val_acc: 0.7472
Epoch 604/1000
713/713 [==============================] - 0s 58us/step - loss: 0.1665 - acc: 0.7714 - val_loss: 0.1388 - val_acc: 0.8034
Epoch 605/1000
713/713 [==============================] - 0s 67us/step - loss: 0.1656 - acc: 0.7714 - val_loss: 0.1370 - val_acc: 0.8146
Epoch 606/1000
713/713 [==============================] - 0s 79us/step - loss: 0.1563 - acc: 0.7924 - val_loss: 0.1833 - val_acc: 0.6910
Epoch 607/1000
713/713 [=================

Epoch 660/1000
713/713 [==============================] - 0s 62us/step - loss: 0.1654 - acc: 0.7728 - val_loss: 0.1322 - val_acc: 0.8258
Epoch 661/1000
713/713 [==============================] - 0s 58us/step - loss: 0.1578 - acc: 0.7882 - val_loss: 0.1320 - val_acc: 0.8258
Epoch 662/1000
713/713 [==============================] - 0s 56us/step - loss: 0.1593 - acc: 0.7784 - val_loss: 0.1495 - val_acc: 0.7978
Epoch 663/1000
713/713 [==============================] - 0s 66us/step - loss: 0.1639 - acc: 0.7714 - val_loss: 0.1311 - val_acc: 0.8315
Epoch 664/1000
713/713 [==============================] - 0s 59us/step - loss: 0.1707 - acc: 0.7644 - val_loss: 0.1334 - val_acc: 0.8315
Epoch 665/1000
713/713 [==============================] - 0s 79us/step - loss: 0.1638 - acc: 0.7672 - val_loss: 0.1351 - val_acc: 0.8315
Epoch 666/1000
713/713 [==============================] - 0s 77us/step - loss: 0.1581 - acc: 0.7882 - val_loss: 0.1423 - val_acc: 0.7921
Epoch 667/1000
713/713 [=================

Epoch 720/1000
713/713 [==============================] - 0s 73us/step - loss: 0.1610 - acc: 0.7714 - val_loss: 0.1466 - val_acc: 0.7809
Epoch 721/1000
713/713 [==============================] - 0s 76us/step - loss: 0.1601 - acc: 0.7854 - val_loss: 0.1424 - val_acc: 0.8034
Epoch 722/1000
713/713 [==============================] - 0s 84us/step - loss: 0.1622 - acc: 0.7602 - val_loss: 0.1550 - val_acc: 0.7921
Epoch 723/1000
713/713 [==============================] - 0s 63us/step - loss: 0.1618 - acc: 0.7798 - val_loss: 0.1555 - val_acc: 0.7584
Epoch 724/1000
713/713 [==============================] - 0s 72us/step - loss: 0.1614 - acc: 0.7658 - val_loss: 0.1315 - val_acc: 0.8315
Epoch 725/1000
713/713 [==============================] - 0s 59us/step - loss: 0.1556 - acc: 0.7952 - val_loss: 0.1328 - val_acc: 0.8258
Epoch 726/1000
713/713 [==============================] - 0s 70us/step - loss: 0.1601 - acc: 0.7756 - val_loss: 0.1369 - val_acc: 0.8258
Epoch 727/1000
713/713 [=================

Epoch 780/1000
713/713 [==============================] - 0s 74us/step - loss: 0.1632 - acc: 0.7742 - val_loss: 0.1666 - val_acc: 0.7191
Epoch 781/1000
713/713 [==============================] - 0s 81us/step - loss: 0.1650 - acc: 0.7672 - val_loss: 0.1508 - val_acc: 0.7640
Epoch 782/1000
713/713 [==============================] - 0s 70us/step - loss: 0.1574 - acc: 0.7882 - val_loss: 0.1497 - val_acc: 0.7921
Epoch 783/1000
713/713 [==============================] - 0s 69us/step - loss: 0.1629 - acc: 0.7798 - val_loss: 0.1430 - val_acc: 0.8034
Epoch 784/1000
713/713 [==============================] - 0s 66us/step - loss: 0.1561 - acc: 0.7896 - val_loss: 0.1756 - val_acc: 0.7865
Epoch 785/1000
713/713 [==============================] - 0s 72us/step - loss: 0.1639 - acc: 0.7714 - val_loss: 0.1404 - val_acc: 0.8090
Epoch 786/1000
713/713 [==============================] - 0s 69us/step - loss: 0.1652 - acc: 0.7630 - val_loss: 0.1326 - val_acc: 0.8427
Epoch 787/1000
713/713 [=================

Epoch 840/1000
713/713 [==============================] - 0s 53us/step - loss: 0.1592 - acc: 0.7826 - val_loss: 0.1498 - val_acc: 0.7978
Epoch 841/1000
713/713 [==============================] - 0s 53us/step - loss: 0.1618 - acc: 0.7728 - val_loss: 0.1339 - val_acc: 0.8146
Epoch 842/1000
713/713 [==============================] - 0s 63us/step - loss: 0.1551 - acc: 0.7868 - val_loss: 0.1778 - val_acc: 0.7809
Epoch 843/1000
713/713 [==============================] - 0s 72us/step - loss: 0.1556 - acc: 0.7826 - val_loss: 0.1643 - val_acc: 0.7865
Epoch 844/1000
713/713 [==============================] - ETA: 0s - loss: 0.1783 - acc: 0.781 - 0s 66us/step - loss: 0.1594 - acc: 0.7700 - val_loss: 0.1443 - val_acc: 0.8090
Epoch 845/1000
713/713 [==============================] - 0s 70us/step - loss: 0.1565 - acc: 0.7798 - val_loss: 0.1284 - val_acc: 0.8258
Epoch 846/1000
713/713 [==============================] - 0s 73us/step - loss: 0.1558 - acc: 0.7784 - val_loss: 0.1376 - val_acc: 0.8258
Epo

713/713 [==============================] - 0s 70us/step - loss: 0.1692 - acc: 0.7588 - val_loss: 0.1299 - val_acc: 0.8034
Epoch 900/1000
713/713 [==============================] - 0s 51us/step - loss: 0.1540 - acc: 0.7784 - val_loss: 0.1305 - val_acc: 0.8258
Epoch 901/1000
713/713 [==============================] - 0s 60us/step - loss: 0.1583 - acc: 0.7812 - val_loss: 0.1316 - val_acc: 0.8090
Epoch 902/1000
713/713 [==============================] - 0s 51us/step - loss: 0.1592 - acc: 0.7742 - val_loss: 0.1282 - val_acc: 0.8258
Epoch 903/1000
713/713 [==============================] - 0s 70us/step - loss: 0.1597 - acc: 0.7756 - val_loss: 0.1272 - val_acc: 0.8371
Epoch 904/1000
713/713 [==============================] - 0s 80us/step - loss: 0.1537 - acc: 0.7896 - val_loss: 0.1717 - val_acc: 0.7079
Epoch 905/1000
713/713 [==============================] - 0s 70us/step - loss: 0.1619 - acc: 0.7700 - val_loss: 0.1323 - val_acc: 0.8090
Epoch 906/1000
713/713 [==============================] 

Epoch 959/1000
713/713 [==============================] - 0s 52us/step - loss: 0.1528 - acc: 0.7826 - val_loss: 0.1626 - val_acc: 0.7921
Epoch 960/1000
713/713 [==============================] - 0s 56us/step - loss: 0.1598 - acc: 0.7742 - val_loss: 0.1342 - val_acc: 0.8371
Epoch 961/1000
713/713 [==============================] - 0s 55us/step - loss: 0.1582 - acc: 0.7630 - val_loss: 0.1954 - val_acc: 0.7584
Epoch 962/1000
713/713 [==============================] - 0s 70us/step - loss: 0.1626 - acc: 0.7756 - val_loss: 0.1272 - val_acc: 0.8427
Epoch 963/1000
713/713 [==============================] - 0s 62us/step - loss: 0.1535 - acc: 0.7938 - val_loss: 0.2346 - val_acc: 0.6629
Epoch 964/1000
713/713 [==============================] - 0s 58us/step - loss: 0.1652 - acc: 0.7616 - val_loss: 0.1269 - val_acc: 0.8315
Epoch 965/1000
713/713 [==============================] - 0s 67us/step - loss: 0.1572 - acc: 0.7770 - val_loss: 0.1278 - val_acc: 0.8315
Epoch 966/1000
713/713 [=================

학습 결과: 조교님 데이터에선 validation accuracy가 최종적으로 72%로 나옴. 중간에 80%정도도 나왔었다가 마지막엔 낮아진 걸로 봐서 overfitting됐다고 볼 수 있음 . training accuracy는 안떨어지는뎅.<br>
-> 중간에 training을 중단시켜주면 됨. 그러나 어느 순간이 최적의 순간인지 알 수 없음.<br>
-> 혹은 regularization 적용.

In [52]:
from keras.regularizers import l2

In [62]:
nn_model3 = Sequential()

In [63]:
nn_model3.add(InputLayer((10,)))

nn_model3.add(Dense(units=10, activation='relu', 
                    kernel_regularizer=l2(), bias_regularizer=l2()))    #regularizer 추가. l2() 안에 가중치를 부여할 수 있음.
nn_model3.add(Dense(units=6, activation='relu', kernel_regularizer=l2(), bias_regularizer=l2()))
nn_model3.add(Dense(units=3, activation='relu', kernel_regularizer=l2(), bias_regularizer=l2()))

nn_model3.add(Dense(units=1, activation='sigmoid'))

nn_model3.compile(optimizer='sgd', loss='mean_squared_error', metrics=['accuracy'])

In [60]:
nn_model3.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_14 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_15 (Dense)             (None, 6)                 66        
_________________________________________________________________
dense_16 (Dense)             (None, 3)                 21        
_________________________________________________________________
dense_17 (Dense)             (None, 1)                 4         
Total params: 201
Trainable params: 201
Non-trainable params: 0
_________________________________________________________________


In [64]:
nn_model3.fit(x=train_x, y=train_y, epochs=200,
             validation_data=(valid_x, valid_y))    #학습 실행

Train on 713 samples, validate on 178 samples
Epoch 1/200
713/713 [==============================] - 2s 3ms/step - loss: 0.4674 - acc: 0.5849 - val_loss: 0.4627 - val_acc: 0.6404
Epoch 2/200
713/713 [==============================] - 0s 136us/step - loss: 0.4618 - acc: 0.6031 - val_loss: 0.4597 - val_acc: 0.6461
Epoch 3/200
713/713 [==============================] - 0s 194us/step - loss: 0.4590 - acc: 0.6087 - val_loss: 0.4577 - val_acc: 0.6404
Epoch 4/200
713/713 [==============================] - 0s 145us/step - loss: 0.4563 - acc: 0.6087 - val_loss: 0.4553 - val_acc: 0.6404
Epoch 5/200
713/713 [==============================] - 0s 159us/step - loss: 0.4539 - acc: 0.6087 - val_loss: 0.4530 - val_acc: 0.6404
Epoch 6/200
713/713 [==============================] - 0s 131us/step - loss: 0.4517 - acc: 0.6143 - val_loss: 0.4513 - val_acc: 0.6517
Epoch 7/200
713/713 [==============================] - 0s 132us/step - loss: 0.4487 - acc: 0.6199 - val_loss: 0.4486 - val_acc: 0.6517
Epoch 8/200

713/713 [==============================] - 0s 83us/step - loss: 0.3459 - acc: 0.6676 - val_loss: 0.3496 - val_acc: 0.6685
Epoch 62/200
713/713 [==============================] - 0s 83us/step - loss: 0.3440 - acc: 0.6620 - val_loss: 0.3468 - val_acc: 0.6629
Epoch 63/200
713/713 [==============================] - 0s 115us/step - loss: 0.3423 - acc: 0.6718 - val_loss: 0.3449 - val_acc: 0.6629
Epoch 64/200
713/713 [==============================] - 0s 88us/step - loss: 0.3409 - acc: 0.6760 - val_loss: 0.3421 - val_acc: 0.6854
Epoch 65/200
713/713 [==============================] - 0s 108us/step - loss: 0.3391 - acc: 0.6718 - val_loss: 0.3396 - val_acc: 0.6854
Epoch 66/200
713/713 [==============================] - 0s 111us/step - loss: 0.3380 - acc: 0.6690 - val_loss: 0.3392 - val_acc: 0.6798
Epoch 67/200
713/713 [==============================] - 0s 105us/step - loss: 0.3368 - acc: 0.6760 - val_loss: 0.3367 - val_acc: 0.6798
Epoch 68/200
713/713 [==============================] - 0s 104us

713/713 [==============================] - 0s 80us/step - loss: 0.2861 - acc: 0.6732 - val_loss: 0.2802 - val_acc: 0.6966
Epoch 122/200
713/713 [==============================] - 0s 76us/step - loss: 0.2854 - acc: 0.6732 - val_loss: 0.2796 - val_acc: 0.6910
Epoch 123/200
713/713 [==============================] - 0s 66us/step - loss: 0.2847 - acc: 0.6760 - val_loss: 0.2796 - val_acc: 0.6910
Epoch 124/200
713/713 [==============================] - 0s 88us/step - loss: 0.2839 - acc: 0.6732 - val_loss: 0.2782 - val_acc: 0.6910
Epoch 125/200
713/713 [==============================] - 0s 83us/step - loss: 0.2833 - acc: 0.6718 - val_loss: 0.2775 - val_acc: 0.6910
Epoch 126/200
713/713 [==============================] - 0s 80us/step - loss: 0.2826 - acc: 0.6732 - val_loss: 0.2779 - val_acc: 0.6966
Epoch 127/200
713/713 [==============================] - 0s 76us/step - loss: 0.2823 - acc: 0.6760 - val_loss: 0.2764 - val_acc: 0.6966
Epoch 128/200
713/713 [==============================] - 0s 72

713/713 [==============================] - 0s 72us/step - loss: 0.2564 - acc: 0.6774 - val_loss: 0.2503 - val_acc: 0.6966
Epoch 182/200
713/713 [==============================] - 0s 66us/step - loss: 0.2560 - acc: 0.6746 - val_loss: 0.2500 - val_acc: 0.6910
Epoch 183/200
713/713 [==============================] - 0s 72us/step - loss: 0.2558 - acc: 0.6746 - val_loss: 0.2493 - val_acc: 0.6966
Epoch 184/200
713/713 [==============================] - 0s 74us/step - loss: 0.2553 - acc: 0.6774 - val_loss: 0.2490 - val_acc: 0.6966
Epoch 185/200
713/713 [==============================] - 0s 72us/step - loss: 0.2549 - acc: 0.6760 - val_loss: 0.2486 - val_acc: 0.6966
Epoch 186/200
713/713 [==============================] - 0s 79us/step - loss: 0.2547 - acc: 0.6746 - val_loss: 0.2483 - val_acc: 0.6966
Epoch 187/200
713/713 [==============================] - 0s 79us/step - loss: 0.2543 - acc: 0.6732 - val_loss: 0.2482 - val_acc: 0.6966
Epoch 188/200
713/713 [==============================] - 0s 77

조교님 모델에선 학습률 더 안좋게 나옴. l2의 값이 너무 커서 underfitting됐다고 보면 됨. 그럴 경우 l값을 줄여주면 됨.<br>
근데 사실 최근엔 regularization 잘 쓰지 않음 dropout만 써도 거의 잘 학습됨.<br>
Dropout은 현재 많이 쓰고 있는 기법.

In [65]:
from keras.layers import Dropout

In [66]:
nn_model4 = Sequential()

nn_model4.add(InputLayer((10,)))

nn_model4.add(Dense(units=10, activation='relu'))
nn_model4.add(Dropout(rate=0.2))   #rate: 몇%만큼 dropout시킬것인가. 여기서는 hidden unit이 10개이므로 2개를 dropuout시키겠다는 말.
#node의 수가 많을 땐 보통 0.5정도로 잡는다고 함.
nn_model4.add(Dense(units=6, activation='relu'))
nn_model4.add(Dropout(rate=0.2))
nn_model4.add(Dense(units=3, activation='relu'))
nn_model4.add(Dropout(rate=0.2))

nn_model4.add(Dense(units=1, activation='sigmoid'))  #ouput layer에는 dropout 적용시키지 않음.

nn_model4.compile(optimizer='sgd', loss='mean_squared_error', metrics=['accuracy'])

In [67]:
nn_model4.fit(x=train_x, y=train_y, epochs=500,
             validation_data=(valid_x, valid_y))    #학습 실행

Train on 713 samples, validate on 178 samples
Epoch 1/500
713/713 [==============================] - 2s 3ms/step - loss: 0.2561 - acc: 0.5975 - val_loss: 0.2130 - val_acc: 0.6461
Epoch 2/500
713/713 [==============================] - 0s 163us/step - loss: 0.2422 - acc: 0.6031 - val_loss: 0.2171 - val_acc: 0.6461
Epoch 3/500
713/713 [==============================] - 0s 143us/step - loss: 0.2429 - acc: 0.5877 - val_loss: 0.2178 - val_acc: 0.6461
Epoch 4/500
713/713 [==============================] - 0s 163us/step - loss: 0.2453 - acc: 0.5891 - val_loss: 0.2139 - val_acc: 0.6461
Epoch 5/500
713/713 [==============================] - 0s 157us/step - loss: 0.2419 - acc: 0.5764 - val_loss: 0.2209 - val_acc: 0.6461
Epoch 6/500
713/713 [==============================] - 0s 163us/step - loss: 0.2569 - acc: 0.5792 - val_loss: 0.2163 - val_acc: 0.6461
Epoch 7/500
713/713 [==============================] - 0s 161us/step - loss: 0.2435 - acc: 0.6073 - val_loss: 0.2142 - val_acc: 0.6461
Epoch 8/500

Epoch 61/500
713/713 [==============================] - 0s 91us/step - loss: 0.2317 - acc: 0.6339 - val_loss: 0.2130 - val_acc: 0.6404
Epoch 62/500
713/713 [==============================] - 0s 94us/step - loss: 0.2321 - acc: 0.6339 - val_loss: 0.2133 - val_acc: 0.6461
Epoch 63/500
713/713 [==============================] - 0s 108us/step - loss: 0.2263 - acc: 0.6269 - val_loss: 0.2183 - val_acc: 0.6404
Epoch 64/500
713/713 [==============================] - 0s 94us/step - loss: 0.2265 - acc: 0.6297 - val_loss: 0.2089 - val_acc: 0.6517
Epoch 65/500
713/713 [==============================] - 0s 91us/step - loss: 0.2322 - acc: 0.6325 - val_loss: 0.2135 - val_acc: 0.6573
Epoch 66/500
713/713 [==============================] - 0s 116us/step - loss: 0.2297 - acc: 0.6269 - val_loss: 0.2133 - val_acc: 0.6573
Epoch 67/500
713/713 [==============================] - 0s 118us/step - loss: 0.2225 - acc: 0.6410 - val_loss: 0.2068 - val_acc: 0.6517
Epoch 68/500
713/713 [==============================

713/713 [==============================] - 0s 83us/step - loss: 0.2143 - acc: 0.6662 - val_loss: 0.2022 - val_acc: 0.6910
Epoch 122/500
713/713 [==============================] - 0s 86us/step - loss: 0.2223 - acc: 0.6732 - val_loss: 0.2035 - val_acc: 0.6798
Epoch 123/500
713/713 [==============================] - 0s 74us/step - loss: 0.2193 - acc: 0.6536 - val_loss: 0.2050 - val_acc: 0.6798
Epoch 124/500
713/713 [==============================] - 0s 77us/step - loss: 0.2144 - acc: 0.6690 - val_loss: 0.2046 - val_acc: 0.6854
Epoch 125/500
713/713 [==============================] - 0s 81us/step - loss: 0.2182 - acc: 0.6634 - val_loss: 0.2015 - val_acc: 0.6854
Epoch 126/500
713/713 [==============================] - 0s 77us/step - loss: 0.2140 - acc: 0.6634 - val_loss: 0.2051 - val_acc: 0.6798
Epoch 127/500
713/713 [==============================] - 0s 76us/step - loss: 0.2208 - acc: 0.6508 - val_loss: 0.2010 - val_acc: 0.6910
Epoch 128/500
713/713 [==============================] - 0s 80

713/713 [==============================] - 0s 77us/step - loss: 0.2205 - acc: 0.6367 - val_loss: 0.1982 - val_acc: 0.6854
Epoch 182/500
713/713 [==============================] - 0s 83us/step - loss: 0.2211 - acc: 0.6508 - val_loss: 0.1983 - val_acc: 0.6910
Epoch 183/500
713/713 [==============================] - 0s 67us/step - loss: 0.2133 - acc: 0.6648 - val_loss: 0.1981 - val_acc: 0.6854
Epoch 184/500
713/713 [==============================] - 0s 63us/step - loss: 0.2217 - acc: 0.6550 - val_loss: 0.1996 - val_acc: 0.6854
Epoch 185/500
713/713 [==============================] - 0s 77us/step - loss: 0.2154 - acc: 0.6676 - val_loss: 0.1992 - val_acc: 0.6910
Epoch 186/500
713/713 [==============================] - 0s 87us/step - loss: 0.2179 - acc: 0.6522 - val_loss: 0.2006 - val_acc: 0.6910
Epoch 187/500
713/713 [==============================] - 0s 73us/step - loss: 0.2169 - acc: 0.6648 - val_loss: 0.2010 - val_acc: 0.6910
Epoch 188/500
713/713 [==============================] - 0s 69

713/713 [==============================] - 0s 63us/step - loss: 0.2177 - acc: 0.6676 - val_loss: 0.2033 - val_acc: 0.6910
Epoch 242/500
713/713 [==============================] - 0s 66us/step - loss: 0.2140 - acc: 0.6676 - val_loss: 0.1958 - val_acc: 0.6910
Epoch 243/500
713/713 [==============================] - 0s 63us/step - loss: 0.2062 - acc: 0.6788 - val_loss: 0.1938 - val_acc: 0.6910
Epoch 244/500
713/713 [==============================] - 0s 74us/step - loss: 0.2074 - acc: 0.6760 - val_loss: 0.1943 - val_acc: 0.6910
Epoch 245/500
713/713 [==============================] - 0s 69us/step - loss: 0.2100 - acc: 0.6634 - val_loss: 0.1941 - val_acc: 0.6854
Epoch 246/500
713/713 [==============================] - 0s 72us/step - loss: 0.2148 - acc: 0.6634 - val_loss: 0.1947 - val_acc: 0.6910
Epoch 247/500
713/713 [==============================] - 0s 70us/step - loss: 0.2142 - acc: 0.6578 - val_loss: 0.1942 - val_acc: 0.6910
Epoch 248/500
713/713 [==============================] - 0s 72

713/713 [==============================] - 0s 62us/step - loss: 0.2105 - acc: 0.6550 - val_loss: 0.1923 - val_acc: 0.6910
Epoch 302/500
713/713 [==============================] - 0s 69us/step - loss: 0.2091 - acc: 0.6746 - val_loss: 0.1925 - val_acc: 0.6854
Epoch 303/500
713/713 [==============================] - 0s 77us/step - loss: 0.2112 - acc: 0.6816 - val_loss: 0.1936 - val_acc: 0.6854
Epoch 304/500
713/713 [==============================] - 0s 101us/step - loss: 0.2055 - acc: 0.6746 - val_loss: 0.1900 - val_acc: 0.6910
Epoch 305/500
713/713 [==============================] - 0s 91us/step - loss: 0.2129 - acc: 0.6620 - val_loss: 0.1912 - val_acc: 0.6854
Epoch 306/500
713/713 [==============================] - 0s 72us/step - loss: 0.2157 - acc: 0.6564 - val_loss: 0.1894 - val_acc: 0.6966
Epoch 307/500
713/713 [==============================] - 0s 69us/step - loss: 0.2113 - acc: 0.6704 - val_loss: 0.1962 - val_acc: 0.6854
Epoch 308/500
713/713 [==============================] - 0s 7

713/713 [==============================] - 0s 63us/step - loss: 0.2129 - acc: 0.6606 - val_loss: 0.1883 - val_acc: 0.7022
Epoch 362/500
713/713 [==============================] - 0s 69us/step - loss: 0.2024 - acc: 0.6872 - val_loss: 0.1867 - val_acc: 0.6966
Epoch 363/500
713/713 [==============================] - 0s 70us/step - loss: 0.2147 - acc: 0.6788 - val_loss: 0.1875 - val_acc: 0.7022
Epoch 364/500
713/713 [==============================] - 0s 97us/step - loss: 0.2099 - acc: 0.6830 - val_loss: 0.1877 - val_acc: 0.6966
Epoch 365/500
713/713 [==============================] - 0s 86us/step - loss: 0.2082 - acc: 0.6900 - val_loss: 0.1877 - val_acc: 0.7022
Epoch 366/500
713/713 [==============================] - 0s 69us/step - loss: 0.2100 - acc: 0.6732 - val_loss: 0.1856 - val_acc: 0.7022
Epoch 367/500
713/713 [==============================] - 0s 70us/step - loss: 0.2124 - acc: 0.6648 - val_loss: 0.1896 - val_acc: 0.7022
Epoch 368/500
713/713 [==============================] - 0s 70

713/713 [==============================] - 0s 70us/step - loss: 0.2071 - acc: 0.6872 - val_loss: 0.1839 - val_acc: 0.7191
Epoch 422/500
713/713 [==============================] - 0s 70us/step - loss: 0.2071 - acc: 0.6788 - val_loss: 0.1854 - val_acc: 0.7247
Epoch 423/500
713/713 [==============================] - 0s 60us/step - loss: 0.2122 - acc: 0.6690 - val_loss: 0.1826 - val_acc: 0.7247
Epoch 424/500
713/713 [==============================] - 0s 80us/step - loss: 0.2070 - acc: 0.6774 - val_loss: 0.1848 - val_acc: 0.7135
Epoch 425/500
713/713 [==============================] - 0s 69us/step - loss: 0.2092 - acc: 0.6718 - val_loss: 0.1835 - val_acc: 0.7135
Epoch 426/500
713/713 [==============================] - 0s 72us/step - loss: 0.2109 - acc: 0.6606 - val_loss: 0.1863 - val_acc: 0.7135
Epoch 427/500
713/713 [==============================] - 0s 69us/step - loss: 0.2081 - acc: 0.6816 - val_loss: 0.1850 - val_acc: 0.7191
Epoch 428/500
713/713 [==============================] - 0s 56

713/713 [==============================] - 0s 66us/step - loss: 0.2140 - acc: 0.6690 - val_loss: 0.1846 - val_acc: 0.7303
Epoch 482/500
713/713 [==============================] - 0s 67us/step - loss: 0.1996 - acc: 0.6830 - val_loss: 0.1842 - val_acc: 0.7303
Epoch 483/500
713/713 [==============================] - 0s 67us/step - loss: 0.2075 - acc: 0.6718 - val_loss: 0.1876 - val_acc: 0.7303
Epoch 484/500
713/713 [==============================] - 0s 63us/step - loss: 0.2086 - acc: 0.6760 - val_loss: 0.1884 - val_acc: 0.7303
Epoch 485/500
713/713 [==============================] - 0s 77us/step - loss: 0.2080 - acc: 0.6858 - val_loss: 0.1878 - val_acc: 0.7303
Epoch 486/500
713/713 [==============================] - 0s 74us/step - loss: 0.2052 - acc: 0.6802 - val_loss: 0.1884 - val_acc: 0.7360
Epoch 487/500
713/713 [==============================] - 0s 70us/step - loss: 0.2040 - acc: 0.6985 - val_loss: 0.1858 - val_acc: 0.7360
Epoch 488/500
713/713 [==============================] - 0s 70

dropout을 쓰면 training과 validation의 갭이 좀 크게 나옴.

In [68]:
nn_model4.evaluate(x=train_x, y=train_y)   #얼마나 제대로 학습이 됐는지 보는 것. 결과에서 앞이 loss, 뒤가 accuracy 정보.

713/713 [==============================] - 0s 69us/step


[0.19092837309394278, 0.7096774193548387]

In [69]:
nn_model4.evaluate(x=valid_x, y=valid_y)

178/178 [==============================] - 0s 101us/step


[0.18023508251383064, 0.7303370773122552]

evaluate해서 나온 결과가 실제 모델의 결과. train 결과와 valid 결과의 값이 좀 더 가까워짐.<br>
마지막으로 optimizer를 바꿔보자.

In [70]:
nn_model5 = Sequential()

nn_model5.add(InputLayer((10,)))

nn_model5.add(Dense(units=10, activation='relu'))
nn_model5.add(Dropout(rate=0.2))   
nn_model5.add(Dense(units=6, activation='relu'))
nn_model5.add(Dropout(rate=0.2))
nn_model5.add(Dense(units=3, activation='relu'))
nn_model5.add(Dropout(rate=0.2))

nn_model5.add(Dense(units=1, activation='sigmoid'))  

nn_model5.compile(optimizer='adam', loss='mean_squared_error', metrics=['accuracy'])   #optimizer만 adam으로 변경

In [72]:
nn_model5.fit(x=train_x, y=train_y, epochs=70, batch_size=16,
             validation_data=(valid_x, valid_y))

Train on 713 samples, validate on 178 samples
Epoch 1/70
713/713 [==============================] - 2s 3ms/step - loss: 0.5121 - acc: 0.4250 - val_loss: 0.6455 - val_acc: 0.3539
Epoch 2/70
713/713 [==============================] - 0s 327us/step - loss: 0.4843 - acc: 0.4460 - val_loss: 0.6338 - val_acc: 0.3539
Epoch 3/70
713/713 [==============================] - 0s 420us/step - loss: 0.4348 - acc: 0.4586 - val_loss: 0.2505 - val_acc: 0.6573
Epoch 4/70
713/713 [==============================] - 0s 348us/step - loss: 0.3143 - acc: 0.5835 - val_loss: 0.2422 - val_acc: 0.6517
Epoch 5/70
713/713 [==============================] - 0s 355us/step - loss: 0.2998 - acc: 0.5736 - val_loss: 0.2406 - val_acc: 0.6461
Epoch 6/70
713/713 [==============================] - 0s 321us/step - loss: 0.2870 - acc: 0.5736 - val_loss: 0.2392 - val_acc: 0.6461
Epoch 7/70
713/713 [==============================] - 0s 345us/step - loss: 0.2620 - acc: 0.5947 - val_loss: 0.2379 - val_acc: 0.6461
Epoch 8/70
713/713

713/713 [==============================] - 0s 189us/step - loss: 0.2380 - acc: 0.6101 - val_loss: 0.2298 - val_acc: 0.6461
Epoch 62/70
713/713 [==============================] - 0s 164us/step - loss: 0.2382 - acc: 0.6073 - val_loss: 0.2298 - val_acc: 0.6461
Epoch 63/70
713/713 [==============================] - 0s 196us/step - loss: 0.2375 - acc: 0.6101 - val_loss: 0.2298 - val_acc: 0.6461
Epoch 64/70
713/713 [==============================] - 0s 157us/step - loss: 0.2373 - acc: 0.6101 - val_loss: 0.2298 - val_acc: 0.6461
Epoch 65/70
713/713 [==============================] - 0s 146us/step - loss: 0.2375 - acc: 0.6129 - val_loss: 0.2298 - val_acc: 0.6461
Epoch 66/70
713/713 [==============================] - 0s 149us/step - loss: 0.2387 - acc: 0.6087 - val_loss: 0.2298 - val_acc: 0.6461
Epoch 67/70
713/713 [==============================] - 0s 184us/step - loss: 0.2402 - acc: 0.6045 - val_loss: 0.2299 - val_acc: 0.6461
Epoch 68/70
713/713 [==============================] - 0s 168us/ste

1 epoch: 전체 데이터를 한 번 학습 시키는 것.<br>
batch: 전체 데이터 중 일부만 가져와서 여러 번 학습 시킴. <br>
예를 들어 전체 데이터 개수가 1000개이고 batch size가 100이라면 100개의 데이터를 10번 학습함. 10번 학습한 것이 1 epoch.<br>
batch_size를 16으로 설정하면 1 epoch마다 약 50번씩 학습하는 것이므로 epoch 수를 좀 줄여줘도 됨.

학습이 좀 더 빠르게 진행됨. 약 200번 정도 학습한 때에서 결과가 고정됨. loss값이 error인데 마지막으로 가면 error값이 증가하는 모습을 보임. overfitting이 일어나고 있다는 의미. 

In [73]:
from keras.optimizers import SGD, Adam

In [74]:
adam_opt = Adam(lr=0.02, beta_1=0.5, beta_2=0.9)   
#변경할 수 있는 option값이 많음. 위에처럼 parameter에서 변경해주면 option값이 그냥 default값으로 적용됨.

만일 ouput이 binary가 아니라 multi일 경우에는 activation을 sigmoid가 아니라 softmax로 설정해줌.<br>
loss도 'categorical_crossentropy'로 설정. binary일 경우에도 binary_crossentropy 사용해도 됨.